manifest.go

```diff
@@ -0,0 +1,655 @@
+	// added to the table. Will be 0 for v1 manifest lists.
+	SequenceNum() int64
+	// MinSequenceNum is the minimum data sequence number of all live data
+	// or delete files in the manifest. Will be 0 for v1 manifest lists.
+	MinSequenceNum() int64
+	// Metadata returns implementation-specific key metadata for encryption
+	// if it exists in the manifest list.
+	Metadata() []byte
+	// PartitionList returns a list of field summaries for each partition
+	// field in the spec. Each field in the list corresponds to a field in
+	// the manifest file's partition spec.
+	PartitionList() []fieldSummary
+
+	// HasAddedFiles returns true if AddedDataFiles > 0 or if it was null.
+	HasAddedFiles() bool
+	// HasExistingFiles returns true if ExistingDataFiles > 0 or if it was null.
+	HasExistingFiles() bool
+	// FetchEntries reads the manifest list file to fetch the list of
+	// manifest entries using the provided file system IO interface.
+	// If discardDeleted is true, entries for files containing deleted rows
+	// will be skipped.
+	FetchEntries(fs iceio.IO, discardDeleted bool) ([]ManifestEntry, error)
+}
+
+// ReadManifestList reads in an avro manifest list file and returns a slice
+// of manifest files or an error if one is encountered.
+func ReadManifestList(in io.Reader) ([]ManifestFile, error) {
+	dec, err := ocf.NewDecoder(in)
+	if err != nil {
+		return nil, err
+	}
+
+	out := make([]ManifestFile, 0)
+
+	for dec.HasNext() {
+		var file ManifestFile
+		if string(dec.Metadata()["format-version"]) == "2" {
+			file = &manifestFileV2{}
+		} else {
+			file = &manifestFileV1{}
+		}
+
+		if err := dec.Decode(file); err != nil {
+			return nil, err
+		}
+		out = append(out, file)
+	}
+
+	return out, dec.Error()
+}
+
+// ManifestEntryStatus defines constants for the entry status of
+// existing, added or deleted.
+type ManifestEntryStatus int8
+
+const (
+	EntryStatusEXISTING ManifestEntryStatus = 0
+	EntryStatusADDED    ManifestEntryStatus = 1
+	EntryStatusDELETED  ManifestEntryStatus = 2
+)
+
+// ManifestEntryContent defines constants for the type of file contents
+// in the file entries. Data, Position based deletes and equality based
+// deletes.
+type ManifestEntryContent int8
+
+const (
+	EntryContentData       ManifestEntryContent = 0
+	EntryContentPosDeletes ManifestEntryContent = 1
+	EntryContentEqDeletes  ManifestEntryContent = 2
+)
+
+// FileFormat defines constants for the format of data files.
+type FileFormat string
+
+const (
+	AvroFile    FileFormat = "AVRO"
+	OrcFile     FileFormat = "ORC"
+	ParquetFile FileFormat = "PARQUET"
+)
+
+type colMap[K, V any] struct {
+	Key   K `avro:"key"`
+	Value V `avro:"value"`
+}
+
+func avroColMapToMap[K comparable, V any](c *[]colMap[K, V]) map[K]V {
+	if c == nil {
+		return nil
+	}
+
+	out := make(map[K]V)
+	for _, data := range *c {
+		out[data.Key] = data.Value
+	}
+	return out
+}
+
+type dataFile struct {
+	Content          ManifestEntryContent   `avro:"content"`
```

Comment:
this only exists for V2 but not for V1

Reply:
Is it a bad thing for it to get written for a v1 file? (readers would just ignore the field, right?)

Reply:
readers would/should ignore it. Would it cause too much trouble to not write this field?

Reply:
In theory, it potentially would avoid writing this field for V1 since it isn't in the schema, but I'll double check

Reply:
Confirmed that if i start with a datafile containing some non-default value for `Content` the round trip test comes back with the default being what the struct has. so it doesn't get written and won't be read back out since the V1 schemas don't have the field.

Reply:
great, thanks for confirming